code/modules/mob/living/carbon/xenomorph/hive_upgrades.dm

```diff
@@ -0,0 +1,177 @@
+	SHOULD_CALL_PARENT(TRUE)
+	if(SSpoints.xeno_points_by_hive[buyer.hivenumber] < psypoint_cost)
+		if(!silent)
+			to_chat(buyer, span_xenowarning("You need [psypoint_cost-SSpoints.xeno_points_by_hive[buyer.hivenumber]] more points to request this boon!"))
+		return FALSE
+	return TRUE
+
+GLOBAL_LIST_INIT(upgrade_categories, list("Buildings", "Defences", "Xenos", "Primordial"))
+/datum/hive_upgrade/building
+	category = "Buildings"
+
+/datum/hive_upgrade/building/silo
+	name = "Larva Silo"
+	desc = "Constructs a silo that generates xeno larvas over time. Requires open space and time to place."
+	psypoint_cost = SILO_PRICE
+	icon = "larvasilo"
+	flags_upgrade = ABILITY_DISTRESS
+
+/datum/hive_upgrade/building/silo/can_buy(mob/living/carbon/xenomorph/buyer, silent = TRUE)
+	. = ..()
+	if(!.)
+		return
+
+	var/turf/buildloc = get_step(buyer, SOUTHWEST)
+	if(!buildloc)
+		return FALSE
+
+	if(buildloc.density)
+		if(!silent)
+			to_chat(buyer, span_xenowarning("You cannot build in a dense location!"))
+		return FALSE
+
+/datum/hive_upgrade/building/silo/on_buy(mob/living/carbon/xenomorph/buyer)
+	if(!do_after(buyer, 10 SECONDS, TRUE, buyer, BUSY_ICON_BUILD))
+		return
+
+	if(!can_buy(buyer, FALSE))
+		return
+
+	var/turf/buildloc = get_step(buyer, SOUTHWEST)
+
+	for(var/obj/structure/xeno/silo/silo AS in GLOB.xeno_resin_silos)
+		if(get_dist(silo, buildloc) < 15)
+			to_chat(buyer, span_xenowarning("Another silo is too close!"))
+			return
+
+	if(SSpoints.xeno_points_by_hive[buyer.hivenumber] < psypoint_cost)
+		to_chat(buyer, span_xenowarning("Someone used all the psych points while we were building!"))
+		return
+
+	to_chat(buyer, span_notice("We build a new silo for [psypoint_cost] psy points."))
+	log_game("[buyer] has built a silo in [AREACOORD(buildloc)], spending [psypoint_cost] psy points in the process")
+	new /obj/structure/xeno/silo(buildloc, buyer.hivenumber)
+	xeno_message("[buyer] has built a silo at [get_area(buildloc)]!", "xenoannounce", 5, buyer.hivenumber)
+	return ..()
+
+/datum/hive_upgrade/defence
+	category = "Defences"
+
+/datum/hive_upgrade/defence/turret
+	name = "Acid turret"
+	desc = "Places a acid spitting resin turret under you. Must be at least 6 tiles away from other turrets, not near fog and on a weeded area."
+	icon = "acidturret"
+	psypoint_cost = XENO_TURRET_PRICE
+	flags_gamemode = ABILITY_DISTRESS
+	var/build_time = 15 SECONDS
+	var/turret_type = /obj/structure/xeno/xeno_turret
+
+/datum/hive_upgrade/defence/turret/can_buy(mob/living/carbon/xenomorph/buyer, silent = TRUE)
+	. = ..()
+	if(!.)
+		return
+	var/turf/T = get_turf(buyer)
+	var/mob/living/carbon/xenomorph/blocker = locate() in T
+	if(blocker && blocker != buyer && blocker.stat != DEAD)
+		if(!silent)
+			to_chat(buyer, span_xenowarning("You cannot build with [blocker] in the way!"))
+		return FALSE
+
+	if(!T.is_weedable())
+		return FALSE
+
+	var/obj/effect/alien/weeds/alien_weeds = locate() in T
+
+	for(var/obj/effect/forcefield/fog/F in range(1, buyer))
```

Comment:
why does this exist?

Plus, you are locating weed and then using it after this line

Reply:
Copy paste code from the ability do i have to change this

Reply:
yeah it's useless